In [ ]:
Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio
test. The function should return the F-value and the corresponding p-value for the test.

In [3]:
import numpy as np
from scipy.stats import f

def var_ratio_test(arr1, arr2):
    #calculate the variances of each array
    var1 = np.var(arr1, ddof=1)
    var2 = np.var(arr2, ddof=1)

    #calculate the F-value
    f_value = var1 / var2

    #calculate the degrees of freedom
    df1 = len(arr1) - 1
    df2 = len(arr2) - 1

    #calculate the p-value
    p_value = f.cdf(f_value, df1, df2)

    return f_value, p_value

arr1 = [1, 2, 3, 4, 5]
arr2 = [6, 7, 8, 9, 10]

f_value, p_value = var_ratio_test(arr1, arr2)

print(f"F-value: {f_value:.2f}")
print(f"P-value: {p_value:.4f}")

F-value: 1.00
P-value: 0.5000


In [ ]:
Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an
F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [4]:
from scipy.stats import f

def critical_f_value(dfn,dfd, alpha=0.05):
    f_crit = f.ppf(1-alpha/2, dfn, dfd)

    return f_crit

dfn = 3
dfd = 16
alpha = 0.05

f_crit = critical_f_value(dfn, dfd, alpha)

print(f"Critical F-value: {f_crit:.4f}")

Critical F-value: 4.0768


In [ ]:
Q3. Write a Python program that generates random samples from two normal distributions with known

variances and uses an F-test to determine if the variances are equal. The program should output the F-
value, degrees of freedom, and p-value for the test.

In [5]:
import numpy as np
from scipy.stats import f

def f_test(sample1, sample2, var1, var2, alpha=0.05):
    s1_var = np.var(sample1, ddof=1)
    s2_var = np.var(sample2, ddof=1)

    f_value = s1_var / s2_var if s1_var > s2_var else s2_var / s1_var

    dfn = len(sample1) - 1 if s1_var > s2_var else len(sample2) - 1
    dfd = len(sample2) - 1 if s1_var > s2_var else len(sample1) - 1

    p_value = f.sf(f_value, dfn, dfd)

    if p_value < alpha:
        print("The null hypothesis that the variances are equal is rejected.")
    else:
        print("The null hypothesis that the variances are equal is not rejected.")

    return f_value, dfn, dfd, p_value


np.random.seed(123)

#generate two random samples from normal distributions with known variances
sample1 = np.random.normal(10, 2, size=20)
sample2 = np.random.normal(10, 3, size=20)

#perform an F-test to determine if the variances are equal
f_value, dfn, dfd, p_value = f_test(sample1, sample2, var1=2**2, var2=3**2)

print(f"F-value: {f_value:.4f}")
print(f"Degrees of freedom (numerator): {dfn}")
print(f"Degrees of freedom (denominator): {dfd}")
print(f"P-value: {p_value:.4f}")

The null hypothesis that the variances are equal is not rejected.
F-value: 1.7070
Degrees of freedom (numerator): 19
Degrees of freedom (denominator): 19
P-value: 0.1264


In [ ]:
Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from
each population. Conduct an F-test at the 5% significance level to determine if the variances are
significantly different.

The null hypothesis is that the variances are equal, and the alternative hypothesis is that they are different.

Here are the steps to conduct the F-test:

Step 1: Calculate the sample variances for each population.

s1^2 = 10
s2^2 = 15

Step 2: Calculate the F-statistic using the formula:

F = s1^2 / s2^2

F = 10 / 15

F = 0.67

Step 3: Calculate the critical value from the F-distribution table.

We have 11 degrees of freedom for each sample (n-1 = 12-1 = 11).

Using a significance level of 0.05 and degrees of freedom 11 and 11, we find the critical value to be 2.818.

Step 4: Compare the F-statistic with the critical value.

Since 0.67 < 2.818, we fail to reject the null hypothesis.

Conclusion:

There is not enough evidence to suggest that the variances of the two populations are significantly different at the 5% significance level.

In [ ]:
Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25
products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance
level to determine if the claim is justified.

Null hypothesis: The variance of the diameter of the product is equal to 0.005.

Alternative hypothesis: The variance of the diameter of the product is greater than 0.005.

Level of significance, α = 0.01

Sample size, n = 25

Sample variance, s^2 = 0.006

Degrees of freedom for the numerator = n - 1 = 25 - 1 = 24

Degrees of freedom for the denominator = n - 1 = 25 - 1 = 24

The F-statistic can be calculated as:

F = (s^2) / (population variance) = 0.006 / 0.005 = 1.2

Using an F-table with degrees of freedom for the numerator = 24 and degrees of freedom for the denominator = 24 and a level of significance of 0.01, the critical value for the F-test is found to be 2.659.

Since the calculated F-value of 1.2 is less than the critical value of 2.659, we fail to reject the null hypothesis. Therefore, there is not enough evidence to conclude that the variance of the diameter of the product is significantly different from 0.005 at the 1% significance level.

In [ ]:
Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an
F-distribution and calculates the mean and variance of the distribution. The function should return the
mean and variance as a tuple.

In [6]:
import numpy as np
from scipy.stats import f

def f_distribution_mean_var(dfn, dfd):
    mean = dfd / (dfd - 2)
    variance = (2 * dfd**2 * (dfn + dfd - 2)) / (dfn * (dfd - 2)**2 * (dfd - 4))
    return (mean, variance)

mean, variance = f_distribution_mean_var(5, 10)
print("Mean: ", mean)
print("Variance: ", variance)

Mean:  1.25
Variance:  1.3541666666666667


In [ ]:
Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The
sample variance is found to be 25. Another random sample of 15 measurements is taken from another
normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test
at the 10% significance level to determine if the variances are significantly different.

To conduct an F-test to determine if the variances of two populations are significantly different, we need to set up the following hypotheses:

Null hypothesis: The variances of the two populations are equal.
Alternative hypothesis: The variances of the two populations are not equal.
We can use the F-test statistic and compare it with the critical value from the F-distribution to make a decision. The F-test statistic is calculated as the ratio of the sample variances, where the larger variance is in the numerator.

In this case, we have:

Sample 1: n1 = 10, s1^2 = 25
Sample 2: n2 = 15, s2^2 = 20
So, the F-test statistic is:

F = s1^2 / s2^2 = 25 / 20 = 1.25

To find the critical value from the F-distribution with 9 and 14 degrees of freedom (10 - 1 and 15 - 1), at the 10% significance level, we can use a calculator or a statistical table. From a statistical table, we find that the critical value is approximately 2.122.

Since our calculated F-test statistic (1.25) is less than the critical value (2.122), we fail to reject the null hypothesis. Therefore, we do not have sufficient evidence to conclude that the variances of the two populations are significantly different at the 10% significance level.

In [9]:
import scipy.stats as stats

df1 = 5
df2 = 5
alpha = 0.05

# Find the critical value from the F table
critical_value = stats.f.ppf(1 - alpha, df1, df2)

print("Critical value from F table:", round(critical_value, 3))

Critical value from F table: 5.05


In [ ]:
Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday
night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5%
significance level to determine if the variances are significantly different.

<img src="Question8.png">

<img src="Question82nd.png">

In [ ]:
Q9. The following data represent the test scores of two groups of students: Group A: 80, 85, 90, 92, 87, 83;
Group B: 75, 78, 82, 79, 81, 84. Conduct an F-test at the 1% significance level to determine if the variances
are significantly different.

<img src="Question9.jpg">

<img src="Question92nd.jpg">